flask/app.py

```diff
@@ -1769,8 +1769,13 @@ def handle_url_build_error(self, error, endpoint, values):
                 rv = handler(error, endpoint, values)
                 if rv is not None:
                     return rv
-            except BuildError as error:
-                pass
+            except BuildError as e:
+                """We can't directly save the BuildError in error because the python3 execution model changed.
+                The `as` target is cleared after exiting the except clause and this causes an UnboundLocalError
+                see  https://docs.python.org/2/reference/compound_stmts.html#except
+                vs.  https://docs.python.org/3.3/reference/compound_stmts.html#except
+                """
+                error = e
```

Comment:
You should set it to `None` outside. Right now it will be undefined if there was no exception and `handler` returned None for some reason.